### ISSUE TITLE
Broken undo after upgrade

### ISSUE THREAD

**Issue by tomitrescak** [xaviergonz/mobx-keystone#110]
Hello, I have upgraded my project from version 0.30.1 to 0.35.1 and many of my tests started to fail because the undo operations were leading to incorrect results. I browsed every version in between mine and the newest one and I found out that undos break when upgrading from `0.32.0` to `0.33.0`.

It is difficult to post an example here, but what I essentially do is that in my @modelAction I update two times a collection holding references to other elements.

First update removes the item from the collection making it empty, second update adds an item to the collection.  When I perform undo, I find the collection to be empty, instead of having one element. This might help you to see what is going on, it is a console log of how pathes are updated:

```
SETTING 
       
      FORWARD PATCH       
      [ { op: 'replace', path: [ 'schemaReferences' ], value: [] } ] 
      at p ​src/@toryjs/form/tests/form_keystone.test.ts:251:6​
       
      INVERSE PATCH      
      [ { op: 'replace', 
          path: [ 'schemaReferences' ], 
          value: [ [Object] ] } ] 
  
     
      FORWARD PATCH        
      [ { op: 'add', 
          path: [ 'schemaReferences', 0 ], 
          value:  
           { elementRef: [Object], 
             propRef: [Object], 
             schemaRef: [Object], 
             '$modelId': '5sk-a8K8bsOpw7sWSgXCqwcIUEQ5wrVZ', 
             '$modelType': 'toryjs/SchemaUseReference' } } ] 
       
      INVERSE PATCH       
      [ { op: 'replace', 
          path: [ 'schemaReferences', 'length' ], 
          value: 0 } ] 

   
      FORWARD PATCH       
      [ { op: 'replace', 
          path:  
           [ 'elementStore', 
             3, 
             'propDataset', 
             'propDataset', 
             'items', 
             'value', 
             'sourceRef' ], 
          value:  
           { id: 'nid', 
             '$modelId': '5sl-a8K8bsOpw7sWSgXCqwcIUEQ5wrVZ', 
             '$modelType': 'toryjs/SchemaElementRef' } } ] 
       
      INVERSE PATCH        
      [ { op: 'replace', 
          path:  
           [ 'elementStore', 
             3, 
             'propDataset', 
             'propDataset', 
             'items', 
             'value', 
             'sourceRef' ], 
          value:  
           { id: 'catId', 
             '$modelId': '5rl-a8K8bsOpw7sWSgXCqwcIUEQ5wrVZ', 
             '$modelType': 'toryjs/SchemaElementRef' } } ] 
       
UNDOING 
       
      FORWARD PATCH        
      [ { op: 'replace', 
          path: [ 'schemaReferences', 0 ], 
          value:  
           { elementRef: [Object], 
             propRef: [Object], 
             schemaRef: [Object], 
             '$modelId': '5rr-a8K8bsOpw7sWSgXCqwcIUEQ5wrVZ', 
             '$modelType': 'toryjs/SchemaUseReference' } } ] 
       
      INVERSE PATCH        
      [ { op: 'replace', 
          path: [ 'schemaReferences', 0 ], 
          value:  
           { elementRef: [Object], 
             propRef: [Object], 
             schemaRef: [Object], 
             '$modelId': '5sk-a8K8bsOpw7sWSgXCqwcIUEQ5wrVZ', 
             '$modelType': 'toryjs/SchemaUseReference' } } ] 
 
      
      FORWARD PATCH        
      [ { op: 'replace', 
          path: [ 'schemaReferences', 'length' ], 
          value: 0 } ] 
      
      INVERSE PATCH        
      [ { op: 'add', 
          path: [ 'schemaReferences', 0 ], 
          value:  
           { elementRef: [Object], 
             propRef: [Object], 
             schemaRef: [Object], 
             '$modelId': '5rr-a8K8bsOpw7sWSgXCqwcIUEQ5wrVZ', 
             '$modelType': 'toryjs/SchemaUseReference' } } ] 
   
    
      FORWARD PATCH       
      [ { op: 'replace', 
          path:  
           [ 'elementStore', 
             3, 
             'propDataset', 
             'propDataset', 
             'items', 
             'value', 
             'sourceRef' ], 
          value:  
           { id: 'catId', 
             '$modelId': '5rl-a8K8bsOpw7sWSgXCqwcIUEQ5wrVZ', 
             '$modelType': 'toryjs/SchemaElementRef' } } ] 
       
      INVERSE PATCH        
      [ { op: 'replace', 
          path:  
           [ 'elementStore', 
             3, 
             'propDataset', 
             'propDataset', 
             'items', 
             'value', 
             'sourceRef' ], 
          value:  
           { id: 'nid', 
             '$modelId': '5sl-a8K8bsOpw7sWSgXCqwcIUEQ5wrVZ', 
             '$modelType': 'toryjs/SchemaElementRef' } } ] 
      at inv ​src/@toryjs/form/tests/form_keystone.test.ts:253:6​
```

**Comment by xaviergonz** [comment]
When you say undo, do you mean you are using the undo manager provided in the library or your own implementation?

If it is your own implementation, are you using applyPatches with reverse set to true when undoing?

If it is the library implementation, any chance you might provide some kind of minimal code reproduction?

From the logs it looks like the inverse patches are getting applied out of order

**Comment by xaviergonz** [comment]
Found it, fixed in v0.37.0, thanks!

**Comment by tomitrescak** [comment]
Thanks! That was fast! I’ll let you know how it went.
